src/main/webapp/app/overview/course-overview.component.ts

```diff
@@ -174,26 +174,63 @@ export class CourseOverviewComponent implements OnInit, OnDestroy, AfterViewInit
+                },
+                error: (res: HttpErrorResponse) => {
+                    if (res.status === 403) {
+                        resolve(false);
+                    } else {
+                        reject(res);
+                    }
+                },
+            });
+        });
+    }
+
+    redirectToCourseRegistrationPage() {
+        this.router.navigate(['courses', this.courseId, 'register']);
+    }
+
     /**
      * Fetch the course from the server including all exercises, lectures, exams and learning goals
      * @param refresh Whether this is a force refresh (displays loader animation)
      */
-    loadCourse(refresh = false) {
+    loadCourse(refresh = false): Promise<void> {
         this.refreshingCourse = refresh;
-        this.courseService.findOneForDashboard(this.courseId, refresh).subscribe({
-            next: (res: HttpResponse<Course>) => {
-                this.courseCalculationService.updateCourse(res.body!);
-                this.course = this.courseCalculationService.getCourse(this.courseId);
-                setTimeout(() => (this.refreshingCourse = false), 500); // ensure min animation duration
-            },
-            error: (error: HttpErrorResponse) => {
-                const errorMessage = error.headers.get('X-artemisApp-message')!;
-                this.alertService.addAlert({
-                    type: AlertType.DANGER,
-                    message: errorMessage,
-                    disableTranslation: true,
-                });
-            },
+        return new Promise<void>((resolve, reject) => {
+            this.courseService.findOneForDashboard(this.courseId, refresh).subscribe({
+                next: (res: HttpResponse<Course>) => {
+                    this.courseCalculationService.updateCourse(res.body!);
+                    this.course = this.courseCalculationService.getCourse(this.courseId);
+                    setTimeout(() => (this.refreshingCourse = false), 500); // ensure min animation duration
+                    resolve();
+                },
+                error: async (error: HttpErrorResponse) => {
+                    if (error.status === 403) {
+                        const canRegister = await this.canRegisterForCourse();
+                        if (canRegister) {
+                            // the user can register, so we want to redirect them
+                            this.redirectToCourseRegistrationPage();
+                            // no need to show the permission error
+                            return;
+                        }
+                    }
+                    const errorMessage = error.headers.get('X-artemisApp-message')!;
+                    this.alertService.addAlert({
+                        type: AlertType.DANGER,
+                        message: errorMessage,
+                        disableTranslation: true,
+                    });
+                    reject();
+                },
+            });
```

Comment:
I did that because I need to `await` it when using it. If I would not await loading the course, some of the other logic (like the metis service) would already fire and produce all kinds of `403` error alerts that I would have to catch / handle somehow.

If you know of a more clever way how I could wait for loading the course and only then do some of the additional logic here, I would be happy to know it!

Reply:
Hmm, interesting point. @krusche, what's your take on that?

Reply:
here is a good explanation: https://stackoverflow.com/a/37365955
(read the whole answer, the part `Observable also has the advantage over Promise to be cancellable` is the important one)

Reply:
Thanks for pointing that out. However, while that's true, do we ever utilise this advantage?
We don't cancel subscriptions, we don't lazyload subscriptions. We do use some operations though.
I think this Person in the comments has a pretty good additional point:

> But please don't forget that Promise, along with async/await makes your code flat again! In a majority of situations, and in projects that don't deal with rocket science, there is no need to write those horrible nested functions with unnecessarily complicated chains of methods. You can use async/await today with transpilers, like TypeScript, and write actual human-readable flat code without any of the rxjs boilerplate. You will probably still need rxjs sometimes in select situations, because it really does have a lot of things to offer. – [evilkos](https://stackoverflow.com/users/1611972/evilkos) [Jul 2, 2017 at 13:59](https://stackoverflow.com/questions/37364973/what-is-the-difference-between-promises-and-observables/37365955#comment76721638_37365955)

So if we utilise or want to utilise the advantages, sure, let's stick with Observables. But if not, we might want to consider slowly migrating.

Wanting to utilise means exactly cancelling all subscriptions on the destruction of every component, I think.